### ISSUE TITLE
When mixed precision input contains grad tensor, FSDP cast it with no grad

### ISSUE THREAD

**Issue by kamwoh** [facebookresearch/fairscale#1191]
https://github.com/facebookresearch/fairscale/blob/5f484b3545f27eddb19d970fbe1d361b9c5f2b07/fairscale/nn/data_parallel/fully_sharded_data_parallel.py#L1424

From this line, notice that it will cast input into appropriate precision with no_grad operation if mixed_precision is on.

In this case, if the input contains grad tensor, for instance, input for this FSDP module is the output computed from another learnable module, then the gradient cannot be backpropagated into this learnable module, causing this learnable module does not receive gradient during optimizer.step()

Wondering why do we casting this with no grad? Is turning off no grad (set to false) safe?

At the moment, i set no_grad as false to bypass the problem.
